qiskit_machine_learning/neural_networks/sampler_qnn.py

```diff
@@ -0,0 +1,343 @@
+
+"""A Neural Network implementation based on the Sampler primitive."""
+
+from __future__ import annotations
+import logging
+
+from numbers import Integral
+from typing import Callable, cast, Iterable, Sequence
+
+import numpy as np
+from qiskit.algorithms.gradients import (
+    BaseSamplerGradient,
+    ParamShiftSamplerGradient,
+    SamplerGradientResult,
+)
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.primitives import BaseSampler, SamplerResult
+from qiskit_machine_learning.exceptions import QiskitMachineLearningError
+import qiskit_machine_learning.optionals as _optionals
+
+from .neural_network import NeuralNetwork
+
+if _optionals.HAS_SPARSE:
+    # pylint: disable=import-error
+    from sparse import SparseArray
+else:
+
+    class SparseArray:  # type: ignore
+        """Empty SparseArray class
+        Replacement if sparse.SparseArray is not present.
+        """
+
+        pass
+
+
+logger = logging.getLogger(__name__)
+
+
+class SamplerQNN(NeuralNetwork):
+    """A Neural Network implementation based on the Sampler primitive."""
+
+    def __init__(
+        self,
+        *,
+        sampler: BaseSampler,
+        circuit: QuantumCircuit,
+        input_params: Sequence[Parameter] | None = None,
+        weight_params: Sequence[Parameter] | None = None,
+        sparse: bool = False,
+        interpret: Callable[[int], int | tuple[int, ...]] | None = None,
+        output_shape: int | tuple[int, ...] | None = None,
+        gradient: BaseSamplerGradient | None = None,
+        input_gradients: bool = False,
+    ):
+        """
+        Args:
+            sampler: The sampler primitive used to compute neural network's results.
+            circuit: The parametrized quantum circuit that generates the samples of this network.
+            input_params: The parameters of the circuit corresponding to the input.
+            weight_params: The parameters of the circuit corresponding to the trainable weights.
+            sparse: Returns whether the output is sparse or not.
+            interpret: A callable that maps the measured integer to another unsigned integer or
+                tuple of unsigned integers. These are used as new indices for the (potentially
+                sparse) output array. If no interpret function is
+                passed, then an identity function will be used by this neural network.
+            output_shape: The output shape of the custom interpretation
+            gradient: An optional sampler gradient to be used for the backward pass.
+            input_gradients: Determines whether to compute gradients with respect to input data.
+                 Note that this parameter is ``False`` by default, and must be explicitly set to
+                 ``True`` for a proper gradient computation when using ``TorchConnector``.
+        Raises:
+            QiskitMachineLearningError: Invalid parameter values.
+        """
+        # set primitive
+        self.sampler = sampler
```

Comment:
For public instance var that we are allowing a user to directly read/update these should be documented in an Attributes: section in the class docstring - here's an example if its needed https://github.com/Qiskit/qiskit-terra/blob/35c4b00c7e9bc142137ee8671be945b910198f96/qiskit/algorithms/time_evolvers/pvqd/pvqd.py#L47-L54